test/scope.coffee

```diff
@@ -33,6 +33,15 @@ test "catch statements should introduce their argument to scope", ->
     do -> e = 5
     eq 5, e
 
+test "#4036 catch statements' error variable should not clash with user-defined variable", ->
+  compile = (code) -> CoffeeScript.compile(code, bare: yes).trim().replace(/\s+/g, " ")
+
+  eq "var _error; try { something(); } catch (__error) { _error = __error; console.log(_error); }", compile("try something()\ncatch _error\n  console.log _error")
+
+  eq "var __error; try { something(); } catch (_error) { __error = _error; console.log(__error); }", compile("try something()\ncatch __error\n  console.log __error")
+
+  eq "var error; try { something(); } catch (_error) { error = _error; console.log(error); }", compile("try something()\ncatch error\n  console.log error")
+
 test "loop variable should be accessible after for-of loop", ->
```

Comment:
Don’t do fragile string comparisons of the output JavaScript. Test that the CoffeeScript works the way intended (just like other tests do). The tests regrading [variables named similar to generated variables](https://github.com/jashkenas/coffeescript/blob/2d1a6fa6ec76b4e721da7e5e4a9c6b90172309ae/test/assignment.coffee#L417-L466) might help.